Test cases

emanuel-skrenkovic opened the pull request:
https://github.com/fsnotify/fsevents/pull/62#issuecomment-2091766682

PR to tests fsnotify style test scripts branched the main branch, as discussed in link above.

emanuel-skrenkovic committed:
copy tests from fsnotify

# Conflicts:
#	fsevents_test.go

emanuel-skrenkovic committed:
try to convert scripts to work with fsevents

emanuel-skrenkovic committed:
always support recurse in tests scripts

emanuel-skrenkovic committed:
remove prints

arp242 commented:
It seems the main bit that's missing here is that it doesn't actually print and parse the list of events? It seems always 0:

            have:
            	0        /
            	0        /file
            want:
            	0        /file

Here's what I would do:

1. Fix that.
2. Just remove all tests, as the behaviours seem too different.
3. Start adding them back one-by-one, fixing up the Output, removing the cross-platform bits that are no longer needed.

You can also use a different approach if that works better for you, but the end result should be something that has tests for at least a bunch of common situations: "make a new directory", "remove watched directory", etc. etc.

emanuel-skrenkovic committed:
remove unsupported

emanuel-skrenkovic committed:
remove not-applicable tests

The tests are based on on the behaviour of a singular watcher across paths. fsevents event streams are simply not used in such a manner - they natively work recursively. There is no need to test for nested path handling.

emanuel-skrenkovic committed:
initial ported tests

emanuel-skrenkovic committed:
initial test script runner code

emanuel-skrenkovic committed:
remove flaky test

The test runs inconsistently. At times fsevents reports all the delete events, at other times it only reports the root directory remove. Removing for now.

emanuel-skrenkovic committed:
remove unused code; make helpers private

emanuel-skrenkovic committed:
remove unnecessary "require recurse"

emanuel-skrenkovic committed:
remove print statement

emanuel-skrenkovic committed:
remove unnecessary "require symlink"

emanuel-skrenkovic committed:
remove commented code

emanuel-skrenkovic committed:
remove other os events from test cases

emanuel-skrenkovic commented:
Added some tests now. The tests that are here run and pass.

Behaviour is different from fsnotify, and I've mostly just replaced the test asserts with what is happening. These aren't really correctness tests at this point, just regression.

I also haven't really had anything to compare against, as there are not too many other tests here. That is something to keep in mind.

I was watching what was happening for each test though, and the events emitted do make sense with regards to what the test does, and what I could find on how fsevents work on mac os.

arp242 commented on helpers_test.go:
```diff
@@ -0,0 +1,729 @@
+	for _, c := range cmds {
+		c := c
+		switch c.cmd {
+		case "skip", "require":
```
I think this command can be removed?

arp242 commented on helpers_test.go:
```diff
@@ -0,0 +1,729 @@
+	"ItemXattrMod":      ItemXattrMod,
+	"ItemIsFile":        ItemIsFile,
+	"ItemIsDir":         ItemIsDir,
+	"ItemIsSymlink":     ItemIsSymlink,
```
Can these just be "created", "removed", "isdir", etc. without the "item" every time? It's very repetitive and "noisy" in the test files.

emanuel-skrenkovic replied:
Hmm, I can change it, it is no problem, but I made it the same as the name of the const to immediately know what is being referred to here. Also makes it grep-able without the extra step. 
It is a bit of a crappy solution admittedly.

Don't have a strong opinion though. If you'd like, I can change it into more human readable names.

arp242 replied:
It's not the biggest thing in the world; you can leave it if that's what you prefer.

arp242 commented on helpers_test.go:
```diff
@@ -0,0 +1,729 @@
+
+		flags := strings.Split(fields[0], "|")
+		for _, f := range flags {
+			resultFlags = resultFlags.set(eventFlagsPossible[f])
```
I don't think this is case-insensitive? It should be: "created", "CREATED", and "Created" should all work. No reason to be pedantic about capitals here.

arp242 commented:
> Behaviour is different from fsnotify, and I've mostly just replaced the test asserts with  what is happening. These aren't really correctness tests at this point, just regression. I also haven't really had anything to compare against, as there are not too many other tests here. That is something to keep in mind.

Yeah, that's fine. The important bit here is to ensure the behaviour doesn't radically change and that it doesn't panic.

Also, can you add a test like so (completely untested code for illustrative purposes):

	func TestMany(t *testing.T) {
		tmp := t.TempDir()
		var wg sync.WaitGroup
		wg.Add(300)
		for i := 0; i<100; i++ {
			i := i
			go func() { defer wg.Done(); echoAppend(fmt.Sprintf(%s/file-%d", tmp, i)) }()
			go func() { defer wg.Done(); mkdir(fmt.Sprintf(%s/dir-%d", tmp, i)) }()
			go func() { defer wg.Done(); echoAppend(fmt.Sprintf(%s/dir-%d/file-%d", tmp, i)) }()
		}
		wg.Wait()

		wg.Add(200)
		for i := 0; i<100; i++ {
			i := i
			go func() { defer wg.Done(); remove(fmt.Sprintf(%s/file-%d", tmp, i)) }()
			go func() { defer wg.Done(); removeAll(fmt.Sprintf(%s/dir-%d", tmp, i)) }()
		}
		wg.Wait()

		// Check if you have 100 writes for new files, 100 removes, etc.
	}


That is, create 100 files and directories simultaneously in a watched directory, then remove them, and ensure we've got the right events.

After that and the minor comments it should be good to go.

emanuel-skrenkovic committed:
add test for many concurrent actions

emanuel-skrenkovic committed:
remove support for skip and remove commands in test scripts

emanuel-skrenkovic committed:
case insensitive flags in test scripts

emanuel-skrenkovic committed:
make TestMany bound by time instead of number of events

Test is being really flaky when using the number of events as the boundary. The number of events emitted is highly variable because the events are batched by the os. In the end, we verify correctness using emitted flags regardless.

emanuel-skrenkovic committed:
short test scripts description in readme

emanuel-skrenkovic committed:
clean up a few comments in helpers_test.go

emanuel-skrenkovic committed:
remove other system references from test scripts

arp242 commented:
Seems good; thanks!
